### PR TITLE
Test signal generator

### LINKED ISSUE #221: Implement a Test Signal Generator Module
**Is your feature request related to a problem? Please describe.**
Implement a test signal generator source module which generates a test pattern for the pipeline.  

**Describe the solution you'd like**
The source process will write a test pattern into a buffer and send in frame.
Eg: A for loop writes a YUV420 test pattern  into a frame. 

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by VishwanathB45**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #221 

**Description**

Test Signal Generator

**Screenshots (if applicable)**

**Checklist**

- [x] I have read the [Contribution Guidelines](https://github.com/Apra-Labs/ApraPipes/wiki/Contribution-Guidelines)
- [x] I have written Unit Tests
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach

### CODE REVIEW

File: base/include/TestSignalGeneratorSrc.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+};
+
+class TestSignalGenerator : public Module
+{
+public:
+    TestSignalGenerator(TestSignalGeneratorProps _props);
+    ~TestSignalGenerator();
+    bool init();
+    bool term();
+
```
Comment (mraduldubey) [inline base/include/TestSignalGeneratorSrc.h:29]: add getProps, setProps methods

File: base/test/testSignalGeneratorSrc_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    }
+    p->run_all_threaded();
+    boost::this_thread::sleep_for(boost::chrono::seconds(1));
+    p->stop();
+    p->term();
+    p->wait_for_all();
+    p.reset();
+}
+
+BOOST_AUTO_TEST_SUITE_END()
```
Comment (mraduldubey) [inline base/test/testSignalGeneratorSrc_tests.cpp:77]: Add one test for dynamic change of properties

**Review by mraduldubey — CHANGES_REQUESTED** [review]
@VishwanathB45 & @Vinayak-YB the tests are not useful at all currently. Please fix them as per suggestions in the review.

File: base/src/TestSignalGeneratorSrc.cpp [hunk 3]
```diff
@@ -85,7 +74,7 @@ bool TestSignalGenerator::init()
 
 bool TestSignalGenerator::produce()
 {
-    size_t read_size = (mDetail->mProps.width * mDetail->mProps.height * 3) >> 1;
+    size_t read_size = (getProps().width * getProps().height * 3) >> 1;
```
Comment (mraduldubey) [inline base/src/TestSignalGeneratorSrc.cpp:77]: This calculation is not required everytime. Only when props are set/changed

File: base/src/TestSignalGeneratorSrc.cpp [hunk 4]
```diff
@@ -106,4 +95,16 @@ void TestSignalGenerator::setMetadata(framemetadata_sp &metadata)
...
+}
+
+TestSignalGeneratorProps TestSignalGenerator::getProps()
+{
+    return mDetail->mProps;
+}
+
+void TestSignalGenerator::setProps(TestSignalGeneratorProps& _props)
+{
+    mDetail->mProps = _props;
```
Comment (mraduldubey) [inline base/src/TestSignalGeneratorSrc.cpp:107]: better to have a Detail::setProps method that is called from here. We might want to do more things in Detail based on props change. Also, the read_size calculation should happen here.

File: base/src/TestSignalGeneratorSrc.cpp (hunk 3 above)
Comment (mraduldubey) [reply base/src/TestSignalGeneratorSrc.cpp:77]: rename read_size to outputFrameSize

File: base/test/testSignalGeneratorSrc_tests.cpp [hunk 5]
```diff
@@ -2,76 +2,88 @@
...
+    }
+};
+
 BOOST_AUTO_TEST_CASE(Basic)
 {
-    auto source = boost::shared_ptr<TestSignalGenerator>(new TestSignalGenerator(TestSignalGeneratorProps(640, 360)));
-    auto metadata = framemetadata_sp(new RawImagePlanarMetadata(640, 360, ImageMetadata::ImageType::YUV420, size_t(0), CV_8U));
+    auto source = boost::shared_ptr<TestSignalGenerator>(new TestSignalGenerator(TestSignalGeneratorProps(400, 400)));
+    auto metadata = framemetadata_sp(new RawImagePlanarMetadata(400, 400, ImageMetadata::ImageType::YUV420, size_t(0), CV_8U));
     source->addOutputPin(metadata);
```
Comment (mraduldubey) [inline base/test/testSignalGeneratorSrc_tests.cpp:40]: addOutputPin is done externally in pipeline in the cases where the module does not know the output pin by itself, for eg. FileReader. That is not the case here. Here, if resolution is given as prop, the output pin is going to always be YUV420 , 8 bit image with same resolution. So, the addOutputPin() should be done inside the module. Not to be done by user in the pipeline.

File: base/test/testSignalGeneratorSrc_tests.cpp [hunk 6]
```diff
@@ -2,76 +2,88 @@
...
-    boost::this_thread::sleep_for(boost::chrono::seconds(2));
-    p->stop();
-    p->term();
-    p->wait_for_all();
-    p.reset();
+    auto frames = sink->try_pop();
+    BOOST_TEST(frames.size() == 1);
+    auto outputFrame = frames.cbegin()->second;
+    BOOST_TEST(outputFrame->getMetadata()->getFrameType() == FrameMetadata::RAW_IMAGE_PLANAR);
+    Test_Utils::saveOrCompare("./data/testOutput/TestSample.raw", const_cast<const uint8_t *>(static_cast<uint8_t *>(outputFrame->data())), outputFrame->size(), 0);
```
Comment (mraduldubey) [inline base/test/testSignalGeneratorSrc_tests.cpp:50]: images not added in PR

File: base/test/testSignalGeneratorSrc_tests.cpp [hunk 7]
```diff
@@ -2,76 +2,88 @@
...
-    boost::this_thread::sleep_for(boost::chrono::seconds(1));
-    p->stop();
-    p->term();
-    p->wait_for_all();
-    p.reset();
+    source->init();
+    sink->init();
+    source->step();
+    auto sinkQue = sink->getQue();
+    auto framedata = sinkQue->pop();
```
Comment (mraduldubey) [inline base/test/testSignalGeneratorSrc_tests.cpp:64]: You should call it frames not framedata. Its a frame_container (with potentially multiple frames). 
Rewrite this without using auto please.

File: base/test/testSignalGeneratorSrc_tests.cpp [hunk 8]
```diff
@@ -2,76 +2,88 @@
...
+    auto frameMetadata = framedata.begin()->second->getMetadata();
+    auto size = frameMetadata->getDataSize() / 1.5;
+    auto currentProps = source->getProps();
+    BOOST_ASSERT(size == currentProps.width * currentProps.height);
+    BOOST_TEST(framedata.size() == 1);
+    auto outputFrame = framedata.cbegin()->second;
+    BOOST_TEST(outputFrame->getMetadata()->getFrameType() == FrameMetadata::RAW_IMAGE_PLANAR);
+    Test_Utils::saveOrCompare("./data/testOutput/TestSample1.raw", const_cast<const uint8_t *>(static_cast<uint8_t *>(outputFrame->data())), outputFrame->size(), 0);
```
Comment (mraduldubey) [inline base/test/testSignalGeneratorSrc_tests.cpp:72]: image not added in PR

File: base/test/testSignalGeneratorSrc_tests.cpp [hunk 9]
```diff
@@ -2,76 +2,88 @@
...
+    TestSignalGeneratorProps propsChange(400, 400);
```
Comment (mraduldubey) [inline base/test/testSignalGeneratorSrc_tests.cpp:73]: call it newProps instead of propsChange

File: base/test/testSignalGeneratorSrc_tests.cpp [hunk 10]
```diff
@@ -2,76 +2,88 @@
...
+    source->setProps(propsChange);
+    source->step();
+    source->step();
```
Comment (mraduldubey) [inline base/test/testSignalGeneratorSrc_tests.cpp:76]: why two steps ? only one should be enough

File: base/test/testSignalGeneratorSrc_tests.cpp [hunk 11]
```diff
@@ -2,76 +2,88 @@
...
+    auto secondque = sink->getQue();
```
Comment (mraduldubey) [inline base/test/testSignalGeneratorSrc_tests.cpp:77]: sinkQue will remain the same. why is secondque variable required ?

File: base/test/testSignalGeneratorSrc_tests.cpp [hunk 12]
```diff
@@ -2,76 +2,88 @@
...
+    auto framedata2 = secondque->pop();
```
Comment (mraduldubey) [inline base/test/testSignalGeneratorSrc_tests.cpp:78]: reuse previous variables

File: base/test/testSignalGeneratorSrc_tests.cpp [hunk 13]
```diff
@@ -2,76 +2,88 @@
...
+    auto frameMetadata2 = framedata2.begin()->second->getMetadata();
+    auto size2 = frameMetadata2->getDataSize() / 1.5;
+    auto currentProps2 = source->getProps();
+    BOOST_ASSERT(size2 == currentProps2.width * currentProps2.height);
+    BOOST_TEST(framedata2.size() == 1);
+    auto outputFrame2 = framedata2.cbegin()->second;
+    BOOST_TEST(outputFrame2->getMetadata()->getFrameType() == FrameMetadata::RAW_IMAGE_PLANAR);
+    Test_Utils::saveOrCompare("./data/testOutput/TestSample2.raw", const_cast<const uint8_t *>(static_cast<uint8_t *>(outputFrame2->data())), outputFrame2->size(), 0);
```
Comment (mraduldubey) [inline base/test/testSignalGeneratorSrc_tests.cpp:86]: image not added in PR

File: base/test/testSignalGeneratorSrc_tests.cpp [hunk 14]
```diff
@@ -2,76 +2,88 @@
...
-    p.reset();
+    auto frames = sink->try_pop();
+    BOOST_TEST(frames.size() == 1);
+    auto outputFrame = frames.cbegin()->second;
+    BOOST_TEST(outputFrame->getMetadata()->getFrameType() == FrameMetadata::RAW_IMAGE_PLANAR);
+    Test_Utils::saveOrCompare("./data/testOutput/TestSample.raw", const_cast<const uint8_t *>(static_cast<uint8_t *>(outputFrame->data())), outputFrame->size(), 0);
 }
 
-BOOST_AUTO_TEST_CASE(fileWriterSink)
+BOOST_AUTO_TEST_CASE(PropsChange)
```
Comment (mraduldubey) [inline base/test/testSignalGeneratorSrc_tests.cpp:53]: call it getSetProps instead

**Review by mraduldubey — CHANGES_REQUESTED** [review]
1. addOutputPin should be internal
2. other minor changes
3. understand what is frame container

File: base/src/TestSignalGeneratorSrc.cpp [hunk 15]
```diff
@@ -97,14 +110,14 @@ void TestSignalGenerator::setMetadata(framemetadata_sp &metadata)
     }
 }
 
-TestSignalGeneratorProps TestSignalGenerator::getProps()
+void TestSignalGenerator::setProps(TestSignalGeneratorProps &_props)
 {
```
Comment (mraduldubey) [inline base/src/TestSignalGeneratorSrc.cpp:114]: this is wrong - use addPropsToQueue and handlePropsChange. we must change the props after the existing frames in queue are processed

**Review by mraduldubey — CHANGES_REQUESTED** [review]
1. setProps flow is incorrect
2. small refactoring

File: base/src/TestSignalGeneratorSrc.cpp (hunk 15 above)
Comment (VishwanathB45) [reply base/src/TestSignalGeneratorSrc.cpp:114]: I have added addPropsToQueue inside the setProps and also created handlePropsChange function.

File: base/src/TestSignalGeneratorSrc.cpp [hunk 16]
```diff
@@ -97,14 +110,14 @@ void TestSignalGenerator::setMetadata(framemetadata_sp &metadata)
...
-    return mDetail->mProps;
+    mDetail->setProps(_props);
+    outputFrameSize = (_props.width * _props.height * 3) >> 1;
+    mDetail->reset();
```
[the remark this answers is not in the document]
Comment (VishwanathB45) [reply base/src/TestSignalGeneratorSrc.cpp:117]: I removed reset from TestSignalGenerator function,added inside mDetail::setProps()

**Review by mraduldubey — DISMISSED** [review]

**Comment by mraduldubey** [comment]
@VishwanathB45 please resolve conflicts and rerun the tests

**Pull request closed without merging**